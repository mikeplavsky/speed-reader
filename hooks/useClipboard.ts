
import {useState, useEffect, useRef} from 'react';
import * as React from 'react';
import {Clipboard, AppState} from 'react-native';

export default function useClipboard() {
  
  const [words, setWords] = useState([]);
  const [idx, setCurrIndex] = useState(0);
  const [text, setText] = useState('');

  const [reading, setReading] = useState(false);
  const id = useRef(0);

  useEffect(() => {

    if (reading) {

      if (idx == words.length - 1) {
        setCurrIndex(0);
      } 

      id.current = setInterval(nextWord,185);

    } 

    return () => {clearInterval(id.current);}; 

  },[reading]);

  useEffect(() => {

    setText(words[0]);

    setReading(false);
    setCurrIndex(0);

  }, [words]);

  useEffect(() => {

    const l = words.length;

    if (idx >= l) {

      setReading(false);
      setCurrIndex(l == 0 ? 0 : l - 1);

      return;
    }

    setText(words[idx]);

  }, [idx]);

  useEffect(() => {

    const stateChange =  (x) => {
      if (x == 'active') { 
        getClipboardText(); 
      }
      else {
        setReading(false);
      }
    };

    AppState.addEventListener("change", stateChange); 
    return () => AppState.removeEventListener("change", stateChange);

  }, []);
  
  const nextWord = () => { 

    setCurrIndex( idx => {
      return idx + 1; 
    }); 

  }
     
  const getClipboardText = async () => {

    const data = await Clipboard.getString();

    setWords(prevWords => {

      if (data.length === 0 ) { return prevWords; }

      let words = data.split(/[\s]+/);
      words = words.filter( x => x.length );
      
      return words;

    });

  }

  const toggleIt = () => {

      setReading( (wasReading) => {
        return !wasReading;
      });

  }

  return [text, toggleIt];

}
