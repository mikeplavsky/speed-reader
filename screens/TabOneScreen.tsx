import * as React from 'react';
import {StyleSheet} from 'react-native';
import {Text, Button} from 'react-native-elements';

import {View} from '../components/Themed';
import App from '../App';

import useClipboard from '../hooks/useClipboard';

export default function TabOneScreen() {

  const [text, toggleIt] = useClipboard();
    
  return (

    <View style={styles.container} 
      onTouchEnd = {toggleIt}
    >
      <Text style={styles.title}>{text}</Text>
    </View>

  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'black',
  },
  title: {
    fontSize: 50,
    color: 'white'
  },
});
