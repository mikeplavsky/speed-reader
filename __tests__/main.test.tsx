import React from 'react';
import {create} from 'react-test-renderer';
import {timer} from '@react-native-community/rxjs';

jest.mock('../hooks/useClipboard');

import useClipboard from "../hooks/useClipboard";
import TabOneScreen from '../screens/TabOneScreen';

afterAll(() => jest.unmock('../hooks/useClipboard'));

describe("TabOnceScreen", () => {

    it ("renders text", () => {
        
        useClipboard.mockImplementation(
            () => ["Once", "toggleItOnce"]);

        expect(create( <TabOneScreen />)) 
            .toMatchSnapshot(); 

        useClipboard.mockImplementation(
            () => ["Twice", "toggleItTwice"]);

        expect(create( <TabOneScreen />)) 
            .toMatchSnapshot(); 

    });

});
