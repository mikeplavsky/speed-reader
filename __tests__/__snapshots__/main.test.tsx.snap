// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabOnceScreen renders text 1`] = `
<View
  onTouchEnd="toggleItOnce"
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "backgroundColor": "black",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Text
    style={
      Object {
        "color": "white",
        "fontSize": 50,
      }
    }
  >
    Once
  </Text>
</View>
`;

exports[`TabOnceScreen renders text 2`] = `
<View
  onTouchEnd="toggleItTwice"
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "backgroundColor": "black",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Text
    style={
      Object {
        "color": "white",
        "fontSize": 50,
      }
    }
  >
    Twice
  </Text>
</View>
`;
